getting_started in Golang

patilsnr opened the pull request:
adds getting_started sample flow in golang using mqtt5

rido-min commented on scenarios/getting_started/go/getting_started.go:
```diff
@@ -0,0 +1,102 @@
+	clientConfig := autopaho.ClientConfig{
+		BrokerUrls: []*url.URL{u},
+		TlsCfg:     cfg,
+		KeepAlive:  30,
```
KeepAlive should be configurable

rido-min reviewed (changes requested):
- This sample should be aligned with other langauges, and refactor common code to mqttclients folder to enable reuse for future samples
- Consistency with default values defined 
- Do not use autopaho
- Update READMEs
- Include instructions for mosquitto (including how to configure the CA)

rido-min commented on scenarios/getting_started/go/getting_started.go:
```diff
@@ -0,0 +1,110 @@
+func main() {
+	// Load connection settings
+	godotenv.Load("../.env")
+	cs := MqttConnectionSettings{
```
we will repeat the same code in all samples. Can you move this to /mqttclients ?

rido-min commented on scenarios/getting_started/go/getting_started.go:
```diff
@@ -0,0 +1,110 @@
+	CaFile    string
+	KeepAlive uint16
+	ClientId  string
+	Username  string
```
why username but no password?

rido-min commented on scenarios/getting_started/go/getting_started.go:
```diff
@@ -0,0 +1,110 @@
+		ClientId:  os.Getenv("MQTT_CLIENT_ID"),
+		CaFile:    os.Getenv("MQTT_CERT_FILE"),
+		KeepAlive: 30,
+		KeyFile:   os.Getenv("MQTT_KEY_FILE"),
```
I miss some initialization, such as UseTls, or tcpPort

rido-min commented on scenarios/getting_started/go/getting_started.go:
```diff
@@ -0,0 +1,110 @@
+
+	// Load certificates
+	fmt.Println("Loading certificates")
+	cert, err := tls.LoadX509KeyPair(fmt.Sprintf("../%s", cs.CaFile), fmt.Sprintf("../%s", cs.KeyFile))
```
mmhhh LoadX509KeyPair seems needed to load the private key, not needed here. Is there any LoadPem? or similar

rido-min commented on scenarios/getting_started/go/getting_started.go:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	// Load certificates
+	fmt.Println("Loading certificates")
```
we should load the CAFile only if the envvar was set

rido-min commented on scenarios/getting_started/go/getting_started.go:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	cfg := &tls.Config{
+		Certificates: []tls.Certificate{cert},
```
we need to load client certificate if specified

rido-min commented on scenarios/getting_started/go/getting_started.go:
```diff
@@ -0,0 +1,110 @@
+		Certificates: []tls.Certificate{cert},
+	}
+
+	fmt.Println("Dialing Eventgrid")
```
Dialing?

rido-min commented on scenarios/getting_started/go/getting_started.go:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	fmt.Println("Dialing Eventgrid")
+	conn, err := tls.Dial("tcp", fmt.Sprintf("%s:%d", cs.Hostname, cs.TcpPort), cfg)
```
what if UseTls was set to false?

rido-min commented on scenarios/getting_started/go/getting_started.go:
```diff
@@ -0,0 +1,110 @@
+	cp := &paho.Connect{
+		KeepAlive:    cs.KeepAlive,
+		ClientID:     cs.ClientId,
+		CleanStart:   true,
```
why not configuring CleanStart ?

rido-min commented on scenarios/getting_started/go/getting_started.go:
```diff
@@ -0,0 +1,110 @@
+		TcpPort:   8883,
+		Username:  os.Getenv("MQTT_USERNAME"),
+		ClientId:  os.Getenv("MQTT_CLIENT_ID"),
+		CaFile:    os.Getenv("MQTT_CERT_FILE"),
```
do not confuse CaFile with CertFile

rido-min reviewed (changes requested):
missing some envvars
move code to mosquitto clients to enable future reuse
Clarify when to use client certs vs. CaFile

Have you tested with Mosquitto ?

We also need to update all the READMEs

rido-min commented on scenarios/getting_started/go/getting_started.go:
```diff
@@ -0,0 +1,101 @@
+func main() {
+	// Load connection settings
+
+	cs := ConnectionSettings.loadConnectionSettings("../.env")
```
we should not hardcode the path to the .env file, or we accept as an argument, or we try to load from current dir.

rido-min reviewed (changes requested):
This is looking great !!, 

although I think it still missing tests, and README updates (not only the Getting Started README).

Have you tested in mosquitto and EG?
Is the codespaces devContainer ready for GO?

rido-min reviewed (changes requested):
any chance to unit test connectionsettings.go?
I'm missing some logic to enable/disable TLS

Have you tried with mosquitto?

patilsnr commented:
> any chance to unit test connectionsettings.go? I'm missing some logic to enable/disable TLS
> 
> Have you tried with mosquitto?

I'm still addressing your comments and pushing periodically. I will let you know when this is ready for a proper review.

rido-min commented:
> > any chance to unit test connectionsettings.go? I'm missing some logic to enable/disable TLS
> > Have you tried with mosquitto?
> 
> I'm still addressing your comments and pushing periodically. I will let you know when this is ready for a proper review.

you can set the PR as draft until it's ready

rido-min commented on .gitignore:
```diff
@@ -8,6 +8,7 @@
 *.pem
 *.crt
 *.env
+*.exe
```
this will only work on Windows, not linux (no extension)

We should find a way to exclude the output folder.

patilsnr commented:
/azp run

rido-min reviewed (approved):
LGTM
